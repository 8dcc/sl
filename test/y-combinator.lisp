;; For more information on the Y combinator, see my blog article:
;; https://8dcc.github.io/programming/understanding-y-combinator.html

(define Y
  (lambda (f)
    ((lambda (x) (f (lambda (n) ((x x) n))))
     (lambda (x) (f (lambda (n) ((x x) n)))))))

(define fact-generator
  (lambda (self)
    (lambda (n)
      (if (equal? n 0)
          1
          (* n (self (- n 1)))))))

(define fact
  (Y fact-generator))

(fact 5)
