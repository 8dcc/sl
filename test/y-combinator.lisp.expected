<lambda>
<lambda>
<lambda>
120
